File: src/types/jit.ts
/**
 * JIT (Just-In-Time) Task Engine Types
 *
 * Event-driven task suggestion system that calculates optimal next tasks
 * on demand rather than pre-computing schedules.
 */

// ─────────────────────────────────────────────────────────────────────────────
// Core Types
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Current context for JIT calculations
 */
export interface JitContext {
	/** Current energy level (0-100) */
	energy: number;
	/** Time since last break (minutes) */
	time_since_last_break_min: number;
	/** Currently running task (if any) */
	current_task: TaskSummary | null;
	/** Number of completed focus sessions today */
	completed_sessions: number;
	/** Current timestamp for context (ISO string) */
	now: string;
}

/**
 * Summary of a task for suggestion purposes
 */
export interface TaskSummary {
	id: string;
	title: string;
	required_minutes: number | null;
	energy: EnergyLevel;
	priority: number;
}

/**
 * Energy level for tasks
 */
export type EnergyLevel = "low" | "medium" | "high";

/**
 * Why this task was suggested
 */
export type SuggestionReason =
	/** Highest priority ready task */
	| "HighPriority"
	/** Matches current energy level */
	| "EnergyMatch"
	/** Quick win (short duration) */
	| "QuickWin"
	/** Most recently deferred */
	| "RecentlyDeferred"
	/** Part of active project */
	| "ActiveProject";

/**
 * Suggested task with priority score
 */
export interface TaskSuggestion {
	task: TaskSummary;
	/** Higher is better (0-100) */
	score: number;
	/** Reason for this suggestion */
	reason: SuggestionReason;
}

// ─────────────────────────────────────────────────────────────────────────────
// Tauri Command Types
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Invoke handler for JIT commands
 */
export interface JitCommands {
	/**
	 * Suggest next tasks based on current context
	 * @param energy Current energy level (0-100)
	 * @param time_since_break Time since last break in minutes
	 * @param completed_sessions Number of completed sessions today
	 * @returns Up to 3 task suggestions, sorted by score
	 */
	jit_suggest_next_tasks: (
		energy?: number,
		time_since_break?: number,
		completed_sessions?: number,
	) => Promise<TaskSuggestion[]>;

	/**
	 * Suggest optimal break duration based on context
	 * @param energy Current energy level (0-100)
	 * @param completed_sessions Number of completed sessions today
	 * @returns Suggested break duration in minutes
	 */
	jit_suggest_break_duration: (
		energy?: number,
		completed_sessions?: number,
	) => Promise<number>;

	/**
	 * Check if user should take a break now
	 * @param energy Current energy level (0-100)
	 * @param time_since_break Time since last break in minutes
	 * @param completed_sessions Number of completed sessions today
	 * @returns true if break is recommended
	 */
	jit_should_take_break: (
		energy?: number,
		time_since_break?: number,
		completed_sessions?: number,
	) => Promise<boolean>;
}

// ─────────────────────────────────────────────────────────────────────────────
// Helper Functions
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Get suggestion reason label for display
 */
export function getSuggestionReasonLabel(reason: SuggestionReason): string {
	switch (reason) {
		case "HighPriority":
			return "High Priority";
		case "EnergyMatch":
			return "Matches Your Energy";
		case "QuickWin":
			return "Quick Win";
		case "RecentlyDeferred":
			return "Previously Deferred";
		case "ActiveProject":
			return "Active Project";
		default:
			return "Suggested";
	}
}

/**
 * Get suggestion reason icon
 */
export function getSuggestionReasonIcon(reason: SuggestionReason): string {
	switch (reason) {
		case "HighPriority":
			return "🔥";
		case "EnergyMatch":
			return "⚡";
		case "QuickWin":
			return "⚡";
		case "RecentlyDeferred":
			return "🔄";
		case "ActiveProject":
			return "📁";
		default:
			return "💡";
	}
}

File: crates/pomodoroom-cli/src/commands/jit.rs
//! JIT (Just-In-Time) task engine command.

use clap::Subcommand;
use chrono::Utc;

use pomodoroom_core::{JitContext, JitEngine};
use pomodoroom_core::storage::schedule_db::ScheduleDb;

#[derive(Subcommand)]
pub enum JitAction {
    /// Suggest next tasks based on current context
    Suggest {
        /// Current energy level (0-100)
        #[arg(long)]
        energy: Option<u8>,
        /// Time since last break in minutes
        #[arg(long)]
        time_since_break: Option<u64>,
        /// Number of completed sessions today
        #[arg(long)]
        completed: Option<u32>,
    },
    /// Suggest optimal break duration
    SuggestBreak {
        /// Current energy level (0-100)
        #[arg(long)]
        energy: Option<u8>,
        /// Number of completed sessions today
        #[arg(long)]
        completed: Option<u32>,
    },
    /// Check if you should take a break
    ShouldBreak {
        /// Current energy level (0-100)
        #[arg(long)]
        energy: Option<u8>,
        /// Time since last break in minutes
        #[arg(long)]
        time_since_break: Option<u64>,
        /// Number of completed sessions today
        #[arg(long)]
        completed: Option<u32>,
    },
}

pub fn run(action: JitAction) -> Result<(), Box<dyn std::error::Error>> {
    match action {
        JitAction::Suggest { energy, time_since_break, completed } => {
            suggest_tasks(energy, time_since_break, completed);
        }
        JitAction::SuggestBreak { energy, completed } => {
            suggest_break(energy, completed);
        }
        JitAction::ShouldBreak { energy, time_since_break, completed } => {
            check_should_break(energy, time_since_break, completed);
        }
    }
    Ok(())
}

fn build_context(
    energy: Option<u8>,
    time_since_break: Option<u64>,
    completed: Option<u32>,
) -> JitContext {
    JitContext {
        energy: energy.unwrap_or(50),
        time_since_last_break_min: time_since_break.unwrap_or(0),
        current_task: None,
        completed_sessions: completed.unwrap_or(0),
        now: Utc::now(),
    }
}

fn suggest_tasks(
    energy: Option<u8>,
    time_since_break: Option<u64>,
    completed: Option<u32>,
) {
    let db = match ScheduleDb::open() {
        Ok(db) => db,
        Err(e) => {
            eprintln!("Error: Could not open database: {}", e);
            return;
        }
    };

    let tasks = match db.list_tasks() {
        Ok(tasks) => tasks,
        Err(e) => {
            eprintln!("Error: Could not list tasks: {}", e);
            return;
        }
    };

    let context = build_context(energy, time_since_break, completed);
    let engine = JitEngine::new();
    let suggestions = engine.suggest_next_tasks(&context, &tasks);

    if suggestions.is_empty() {
        println!("No task suggestions available.");
        println!("  Make sure you have READY tasks in your Active list.");
        return;
    }

    println!("=== Task Suggestions ===\n");
    for (i, suggestion) in suggestions.iter().enumerate() {
        println!("{}. {} (Score: {})",
            i + 1,
            suggestion.task.title,
            suggestion.score
        );
        println!("   Reason: {:?}", suggestion.reason);
        if let Some(duration) = suggestion.task.required_minutes {
            println!("   Duration: {} min", duration);
        }
        println!("   Priority: {}", suggestion.task.priority);
        println!();
    }

    // Show context summary
    println!("Context:");
    println!("  Energy: {}/100", context.energy);
    println!("  Time since break: {} min", context.time_since_last_break_min);
    println!("  Completed sessions: {}", context.completed_sessions);
}

fn suggest_break(energy: Option<u8>, completed: Option<u32>) {
    let context = build_context(energy, None, completed);
    let engine = JitEngine::new();
    let duration = engine.suggest_break_duration(&context);

    println!("=== Break Suggestion ===");
    println!("  Recommended duration: {} minutes", duration);
    println!("  Current energy: {}/100", context.energy);
    println!("  Completed sessions: {}", context.completed_sessions);

    if duration >= 15 {
        println!("\n  Tip: This is a long break. Consider stepping away from the screen.");
    }
}

fn check_should_break(
    energy: Option<u8>,
    time_since_break: Option<u64>,
    completed: Option<u32>,
) {
    let context = build_context(energy, time_since_break, completed);
    let engine = JitEngine::new();
    let should_break = engine.should_take_break(&context);

    println!("=== Break Check ===");
    if should_break {
        println!("  ⚠️  Yes, you should take a break!");
        println!("  Current energy: {}/100", context.energy);
        println!("  Time since break: {} min", context.time_since_last_break_min);

        let duration = engine.suggest_break_duration(&context);
        println!("  Suggested duration: {} min", duration);
    } else {
        println!("  ✓ No, keep going!");
        println!("  Your energy is still good ({}/100)", context.energy);
    }
}
